File: src/bin/cli/cmd/compare.rs
use crate::cli::OutputFormat;
use crate::json_output::output_error;
use crate::utils;
use lib::subtitle::model::Subtitle;
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use log::{debug, info};

pub fn handle(file1: &str, file2: &str, format: &OutputFormat) -> anyhow::Result<()> {
    // Compare is TUI-only, reject JSON output
    if matches!(format, OutputFormat::Json) {
        output_error(
            format,
            "unsupported_format",
            "The compare command is TUI-only and does not support --output json",
            Some("Use 'sm verify' for machine-readable comparison output"),
        );
        return Err(anyhow::anyhow!(""));
    }

    info!("comparing files: {} and {}", file1, file2);

    let (subtitles1, filename1) = load_subtitle_file(file1, format)?;
    let (subtitles2, filename2) = load_subtitle_file(file2, format)?;

    debug!(
        "loaded {} and {} subtitles",
        subtitles1.len(),
        subtitles2.len()
    );

    crate::output::compare::run_tui(subtitles1, filename1, subtitles2, filename2)?;

    info!("comparison completed");
    Ok(())
}

fn load_subtitle_file(
    file: &str,
    format: &OutputFormat,
) -> anyhow::Result<(Vec<Subtitle>, String)> {
    debug!("loading subtitle file: {}", file);

    let resolved = utils::resolve_existing_path(file)?;
    let service = SubRipService::new(resolved.base_dir);

    match service.get_all(&resolved.filename) {
        Ok(subtitles) => {
            info!(
                "successfully loaded {} subtitle(s) from {}",
                subtitles.len(),
                resolved.filename
            );
            Ok((subtitles, resolved.filename))
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

File: internal/cli/cmd/compare.go
// Package cmd holds the command handlers for the submarine CLI. Each handler is
// a 1-to-1 port of the corresponding Rust `src/bin/cli/cmd/*.rs` module.
package cmd

import (
	"errors"
	"fmt"
	"log/slog"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/internal/cli/output"
	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// HandleCompare runs the TUI comparison view for two subtitle files. Compare is
// TUI-only and rejects JSON output. Port of the Rust `cmd::compare::handle`.
func HandleCompare(file1, file2 string, format cli.OutputFormat) error {
	// Compare is TUI-only, reject JSON output
	if format == cli.OutputFormatJson {
		cli.OutputError(
			format,
			"unsupported_format",
			"The compare command is TUI-only and does not support --output json",
			strPtr("Use 'sm verify' for machine-readable comparison output"),
		)
		return errors.New("")
	}

	slog.Info(fmt.Sprintf("comparing files: %s and %s", file1, file2))

	subtitles1, filename1, err := loadSubtitleFile(file1, format)
	if err != nil {
		return err
	}
	subtitles2, filename2, err := loadSubtitleFile(file2, format)
	if err != nil {
		return err
	}

	slog.Debug(fmt.Sprintf("loaded %d and %d subtitles", len(subtitles1), len(subtitles2)))

	if err := output.RunTUI(subtitles1, filename1, subtitles2, filename2); err != nil {
		return err
	}

	slog.Info("comparison completed")
	return nil
}

// loadSubtitleFile loads all subtitles from a file via the SubRip service. On
// error it prints a formatted error and returns an empty-message error to
// signal an already-reported failure. Port of the Rust `load_subtitle_file`.
func loadSubtitleFile(file string, format cli.OutputFormat) ([]subtitle.Subtitle, string, error) {
	slog.Debug(fmt.Sprintf("loading subtitle file: %s", file))

	resolved, err := cli.ResolveExistingPath(file)
	if err != nil {
		return nil, "", err
	}

	service := subtitle.NewSubRipService(resolved.BaseDir)

	subtitles, err := service.GetAll(resolved.Filename)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, file)
			cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
		}
		return nil, "", errors.New("")
	}

	slog.Info(fmt.Sprintf("successfully loaded %d subtitle(s) from %s", len(subtitles), resolved.Filename))
	return subtitles, resolved.Filename, nil
}

// strPtr returns a pointer to s (helper for optional hints).
func strPtr(s string) *string { return &s }
